client/hotrod-client/src/main/java/org/infinispan/client/hotrod/DataFormat.java

```diff
@@ -141,27 +208,43 @@ public byte[] valueToBytes(Object value, int estimateKeySize, int estimateValueS
    }
 
    public byte[] valueToBytes(Object value) {
-      Marshaller valueMarshaller = resolveValueMarshaller();
-      return obj2bytes(valueMarshaller, value, valueSizePredictor);
+      return client.valueToBytes(value);
    }
 
    public <T> T keyToObj(byte[] bytes, ClassAllowList allowList) {
-      Marshaller keyMarshaller = resolveKeyMarshaller();
-      return bytes2obj(keyMarshaller, bytes, isObjectStorage, allowList);
+      return client.bytesToKey(bytes, allowList);
    }
 
    public <T> T valueToObj(byte[] bytes, ClassAllowList allowList) {
-      Marshaller valueMarshaller = resolveValueMarshaller();
-      return bytes2obj(valueMarshaller, bytes, isObjectStorage, allowList);
+      return client.bytesToValue(bytes, allowList);
+   }
+
+   public MediaTypeMarshaller server() {
+      if (server == null || server.match(client)) return null;
+
+      // If is an object storage, we hash the actual object to retrieve the segment.
+      // We return null so the client does not use anything from the server.
+      if (isObjectStorage()) return null;
+
+      // This means that the client DOES NOT have a marshaller capable of converting to the server key type.
+      // Therefore, it will utilize the default fallback and NOT convert the object.
+      // This could cause additional redirections on the server side and poor performance for the client.
+      if (marshallerRegistry.getMarshaller(server.keyType) == null) {
+         log.serverKeyTypeNotRecognized(server.keyType);
```

Comment:
We should be able to move this into the resolveStorage method, which is invoked once per cache. Then the @Once annotation can be removed.

Reply:
Thanks, updated.